### DISCUSSION TITLE
PDF file size is increasing after adding highlight annot drastically

### DISCUSSION THREAD

**Discussion by kvrameshreddy** [pymupdf/PyMuPDF#1623] [Q&A] [answered]
Hi @JorjMcKie,

I have a PDF file of 12kb with a table in it, I need to apply annotations for each word present on it, after saving the file, the size is increasing to 288kb (24 times to actual ). Is there a way to reduce the size of the file? (a file of 400 kb is increasing to 10MB)

attached is the pdf file and code which I used.

```
import os
import fitz
import difflib
pdf1=fitz.open(r"D:\logs\3.pdf")

for i in range(len(pdf1)):
    page1 = pdf1[i]
    new_text1 = page1.get_text("words")
    for j in range(len(new_text1)):
        page1.add_highlight_annot(new_text1[j][0:4])
pdf1.save('D:\logs\check_3.pdf',garbage=3,deflate=True)
```
[3.pdf](https://github.com/pymupdf/PyMuPDF/files/8185111/3.pdf)

Thankyou.

**Comment by JorjMcKie** [comment]
I see no problem given you are adding massive amounts of annotations. Here is your script slightly modified with giving annot numbers. The file size increased from 12 KB to 288 KB. Given a total number of 475 annotations, this means less than 600 bytes per annot.
No issue here.
```python
import fitz

pdf1 = fitz.open("3.pdf")

for page1 in pdf1:
    words = page1.get_text("words")
    for word in words:
        page1.add_highlight_annot(word[:4])
    print(f"Added {len(words)} annots on page {page1.number}.")
pdf1.save("3-out.pdf", garbage=3, deflate=True)
```
Log:
```cmd
py test.py
Added 352 annots on page 0.
Added 123 annots on page 1.
```

**Comment by JorjMcKie** [comment]
The only option you have is **_not adding annots_** but drawing semi-transparent rectangles over the text (or underneath).
These would be permanent changes however - in contrast to annots, which you can remove again.
```python
import fitz

pdf1 = fitz.open("3.pdf")

for page1 in pdf1:
    shape = page1.new_shape()
    words = page1.get_text("words")
    for word in words:
        shape.draw_rect(word[:4])
    shape.finish(fill=(1, 1, 0), fill_opacity=0.3)
    shape.commit()
    print(f"Added {len(words)} rectangles on page {page1.number}.")
pdf1.save("3-out2.pdf", garbage=3, deflate=True)
```

**Reply by JorjMcKie** [reply]
This gives you a new PDF size of only 15 KB. Just a small bit large than the original.

**Reply by kvrameshreddy** [reply]
Thank you @JorjMcKie , this option worked.

**Comment by JorjMcKie** [comment, marked as answer]
Ah, I should have thought of this earlier 😒
You can add **_one_** highlight for any number of rectangles like this:
```python
import fitz

pdf1 = fitz.open("3.pdf")

for page1 in pdf1:
    words = page1.get_text("words")
    word_rects = []
    for word in words:
        word_rects.append(fitz.Rect(word[:4]))
    if word_rects != []:
        page1.add_highlight_annot(word_rects)
    print(f"Added {len(words)} annots on page {page1.number}.")
pdf1.save("3-out.pdf", garbage=3, deflate=True)
```
This makes a 60 KB file from the original 12 KB. Each page only has one annot - with quads as many as there are words on the page.

**Reply by kvrameshreddy** [reply]
Hi @JorjMcKie , thankyou for your immediate response. the two methods working fine.

I have a small question, when we convert each pdf page to image,  and create a separate pdf with images, can we limit the size of the new file same as above

**Reply by JorjMcKie** [reply]
No, has nothing to do with the above.
The output size in that case would depend on the chosen resolution of each page pixmap. Will however probably be larger than the original.
Some size can also be saved if making gray pixmaps - sufficient for text-oriented PDFs.

**Reply by kvrameshreddy** [reply]
Thankyou @JorjMcKie
